### PR TITLE
remove per-feature logging to improve performance issue#685

### PR DESCRIPTION
**Pull request by snowman2**
I saw the issue #685, and so I took a stab at it based on the discussion to see if I could help move things forward.

I commented out the debug print statements instead of deleting them so that you don't have to re-create them when you are debugging (and an official decision on what to do with them has not been reached as far as I know). However, that presents the danger of being re-introduced by accident when committing changes after debugging.

Using this example:
```python
from itertools import chain, repeat
import fiona
def test_to_file():
    with fiona.Env(),fiona.open("../tests/data/coutwildrnp.shp") as collection:
        features = chain.from_iterable(repeat(list(collection), 2000))
        with fiona.open("/tmp/out.gpkg", "w", schema=collection.schema, crs=collection.crs, driver="GPKG") as dst:
            dst.writerecords(features)
```
And timing it:
```python
%%timeit
test_to_file()
```

Before:
1min 3s ± 2.32 s per loop (mean ± std. dev. of 7 runs, 1 loop each)
After:
15.3 s ± 1.13 s per loop (mean ± std. dev. of 7 runs, 1 loop each)

### CODE REVIEW

**Comment by snowman2** [comment]
Looks like the GDAL trunk is using PROJ.6 now: `configure: error: PROJ 6 symbols not found`

**Comment by sgillies** [comment]
@snowman2 wow, great! I think I'll be able to have some review time tomorrow.

**Comment by snowman2** [comment]
Another idea I had recently was change:
https://github.com/Toblerity/Fiona/blob/013dec95f3a468459a76d57b60f676b415bda738/fiona/ogrext.pyx#L320

To instead iterate over:
```python
collection.schema['properties']
```
And set the property to use:
```python
set_field_null(cogr_feature, i)
```
If the property does not exist in `feature['properties']`.

This would eliminate the need for this check:

https://github.com/Toblerity/Fiona/blob/013dec95f3a468459a76d57b60f676b415bda738/fiona/ogrext.pyx#L1167

But, I wanted to run the idea by you first.

File: fiona/_geometry.pyx [hunk 1]
```diff
@@ -283,7 +283,7 @@ cdef class OGRGeomBuilder:
             coordinates = geometry.get('geometries')
             return self._buildGeometryCollection(coordinates)
         else:
-            raise ValueError("Unsupported geometry type %s" % typename)
+            raise UnsupportedGeometryTypeError("Unsupported geometry type %s" % typename)
```
Comment (sgillies) [inline fiona/_geometry.pyx]: 👍

File: fiona/ogrext.pyx [hunk 2]
```diff
@@ -318,19 +318,14 @@ cdef class OGRFeatureBuilder:
...
 
-            log.debug("Normalizing schema type for key %r in schema %r to %r", key, collection.schema['properties'], schema_type)
-
-            try:
-                key_bytes = ogr_key.encode(encoding)
-            except UnicodeDecodeError:
-                log.warning("Failed to encode %s using %s codec", key, encoding)
-                key_bytes = ogr_key
+            # log.debug("Normalizing schema type for key %r in schema %r to %r", key, collection.schema['properties'], schema_type)
+            key_bytes = strencode(ogr_key, encoding)
```
Comment (sgillies) [inline fiona/ogrext.pyx]: 👍

File: fiona/ogrext.pyx [hunk 3]
```diff
@@ -1162,23 +1144,22 @@ cdef class WritingSession(Session):
...
-
+                        record['geometry']['type'],
+                        collection.schema['geometry'] ))
+            # log.debug("Creating feature in layer: %s" % record)
             cogr_feature = OGRFeatureBuilder().build(record, collection)
             result = OGR_L_CreateFeature(cogr_layer, cogr_feature)
             if result != OGRERR_NONE:
+                # Validate against collection's schema to give useful message
+                if set(record['properties'].keys()) != schema_props_keys:
+                    raise ValueError(
```
Comment (sgillies) [inline fiona/ogrext.pyx]: Maybe raise SchemaError here?

**Comment by sgillies** [comment]
@snowman2 what would you think about holding off on the schema/property checks for now? I think it's separable from elimination of the debug statements, which is a big win by itself.

I'm inclined to remove the debug statements that you've commented. I put them in when building the original feature and I don't think they are necessarily needed anymore.

**Comment by snowman2** [comment]
> I'm inclined to remove the debug statements that you've commented.

I removed those in a separate commit.

> what would you think about holding off on the schema/property checks for now?

I checked the difference between checking the properties and the alternative method and I was surprised that it didn't make a significant difference. So, I agree that it is an unnecessary change. It makes me wonder if it would be a good idea to do a check and remove all unnecessary log.debug/log.info statements in rasterio & fiona for performance reasons?

**Comment by sgillies** [comment]
@snorfalorpagus I tagged you for review of this, too.

**Comment by sgillies** [comment]
I edited the PR title to reflect our decision to take a one thing at a time approach to optimization.

**Pull request merged**